### ISSUE TITLE
`Sign in with Apple` option for produce command

### ISSUE THREAD

**Issue by yuzushioh** [fastlane/fastlane#16233]
### Feature Request

Add `Sign in with Apple` option to `Produce`

#### Motivation Behind Feature

I am trying to automate bootstrapping xcode projects for my company. I use `produce` command for generating App IDs and enabling push notification and associated domains feature. But I want to enable `Sign in with Apple` as well.

#### Feature Description

Add `--sign-in-with-apple` to produce's enable_services command.

**Comment by tmaly1980** [comment]
I complained about this back in November, nobody did anything about it. Now gym is failing because Apple demands it currently.

**Comment by yuzushioh** [comment]
@janpio Thanks I will try it 👍🏻

**Comment by tmaly1980** [comment]
@janpio I would love to help implement this, I'm just unsure how. I've looked in the AppService code in Spaceship (which seems to connect to a REST API on Apple Developer's servers), but I cannot find an API reference anywhere online. I've searched for "updateService.action" and the only results point to fastlane issues. 

I'd think this page would be the right documentation, but it's referring to a totally different API, ie api.appstoreconnect.apple.com rather than developer.apple.com/service-account

https://developer.apple.com/documentation/appstoreconnectapi/enable_a_capability

**Comment by tmaly1980** [comment]
@sogko I noticed you were able to find the featureType for some of the app capabilities here:

https://github.com/fastlane/fastlane/issues/5100

How did you figure out what the featureType should be for each capability? They don't seem documented anywhere online. I'm trying to add Apple Sign-In support and don't know how to implement that, since no API seems to be documented.

**Comment by janpio** [comment]
@tmaly1980 Welcome to API hell. Apple is very fond of it.

It is very possible that the _real_ API does not have support for this as you said.
A good first step would be to document in detail what the website does (Screenshots and requests (host, endpoint, params, request and response object - the usual). We might already bee using some of those APIs or at least similar ones, then it is "just" about adding these additional endpoints and building up from there.

**Comment by mattjensjensencs** [comment]
Just echoing the question @tmaly1980 posed,
I spent some time trying to work on it and for the life of me couldn't find where/how the `AppService.service_id`s are being discovered.
I see commits about the `service_id` changing and being updated but don't see the documentation of how to find the IDs in Apple.
The most recent one changed is `Wallet = AppService.new_service("passbook")`, @nafu if you could shed some light on how you found the id it would be greatly appreciated.

**Comment by MartinGonzalez** [comment]
Any news about this. Is there any clue how can we contribute on this?

**Comment by mattjensjensencs** [comment]
So upon investigation I saw that apple returns enabled services with the service ids in the response for `identifiers/getAppIdDetail.action`. However unlike passbook or wireless accessory config no service id appears to be shown for sign in with apple. There's other info where it's enabled if you search for `APPLE_ID_AUTH` which is how I see it referenced on the web requests as well. I'm not very familiar with the internals of the spaceship client but this doesn't seem like adding support for this will be as easy as just finding the service id and adding it to `app_service.rb`.

**Comment by SiberianCMS** [comment]
I've came to the same conclusion, and we need to have the Apple Sign In too!

The payload for `apple_id_auth` is weird compared to other, but well, we should be able to find how and create some pull request

The payload from when I save the "Apple Sign In" from the appstore.

```json
{
  "data": {
    "type": "bundleIds",
    "id": "ZZ1234Z1ZZ",
    "attributes": {
      "identifier": "com.myhidden.bundleid",
      "permissions": {
        "edit": true,
        "delete": true
      },
      "seedId": "TEAM123456",
      "name": "super_app_name super_uuid_hidden",
      "wildcard": false,
      "teamId": "TEAM123456"
    },
    "relationships": {
      "bundleIdCapabilities": {
        "data": [
          {
            "type": "bundleIdCapabilities",
            "attributes": {
              "enabled": true,
              "settings": [
                {
                  "key": "GAME_CENTER_SETTING",
                  "options": [
                    {
                      "key": "GAME_CENTER_IOS"
                    },
                    {
                      "key": "GAME_CENTER_MAC"
                    }
                  ]
                }
              ]
            },
            "relationships": {
              "capability": {
                "data": {
                  "type": "capabilities",
                  "id": "GAME_CENTER"
                }
              }
            }
          },
          {
            "type": "bundleIdCapabilities",
            "attributes": {
              "enabled": true,
              "settings": []
            },
            "relationships": {
              "capability": {
                "data": {
                  "type": "capabilities",
                  "id": "PUSH_NOTIFICATIONS"
                }
              }
            }
          },
          {
            "type": "bundleIdCapabilities",
            "attributes": {
              "enabled": true,
              "settings": []
            },
            "relationships": {
              "capability": {
                "data": {
                  "type": "capabilities",
                  "id": "IN_APP_PURCHASE"
                }
              }
            }
          },
          {
            "type": "bundleIdCapabilities",
            "attributes": {
              "enabled": true,
              "settings": [
                {
                  "key": "TIBURON_APP_CONSENT",
                  "options": [
                    {
                      "key": "PRIMARY_APP_CONSENT"
                    }
                  ]
                }
              ]
            },
            "relationships": {
              "capability": {
                "data": {
                  "type": "capabilities",
                  "id": "APPLE_ID_AUTH"
                }
              }
            }
          }
        ]
      }
    }
  }
}
```


And the extracted part for Apple_Id_Auth

```json
{
  "type": "bundleIdCapabilities",
  "attributes": {
    "enabled": true,
    "settings": [
    {
      "key": "TIBURON_APP_CONSENT",
     "options": [
       {
         "key": "PRIMARY_APP_CONSENT"
        }
      ]
    }
  ] 
},
  "relationships": 
  "capability": {
    "data": {
                  "type": "capabilities",
                  "id": "APPLE_ID_AUTH"
                }
              }
            }
          }
```

So ... well.

**Comment by SiberianCMS** [comment]
Seeing this page

https://developer.apple.com/documentation/appstoreconnectapi/capabilitytype

And the previous payload, I tend to think that the "APPLE_ID_AUTH" is either not yet available from the REST API, or it is a special type of capability that we must enable through something else.

**Comment by jjochen** [comment]
Any news?

**Comment by Nightstep** [comment]
Hi, Any update here?

**Comment by tmaly1980** [comment]
Has anyone made any progress with this?

**Comment by tmaly1980** [comment]
@SiberianCMS Did you ever get anywhere with this?

**Comment by jjochen** [comment]
> We've got white labeled apps that depend upon fastlane and need a way to automate enabling sign in with apple.

Same here

**Comment by Panajev** [comment]
@joshdholtz right now this is auto-closed, but is the project taking this into a future roadmap or is it a consideration for a "Won't Do" list which I think is helpful like the Swift project's commonly rejected proposals list? Seriously, fine with either approach, not trying to put pressure which already abounds :).

**Comment by tmaly1980** [comment]
Besides simply enabling Sign in With Apple for an app (which I believe can be done through the new AppStoreConnect API), what about the private key generation required for verifying the user's token?

https://github.com/invertase/react-native-apple-authentication/blob/master/docs/INITIAL_SETUP.md

We're specifically using white labeled apps via one developer account part of many developer teams, and it would seem like this can't be programmatically done?

**Comment by rogerluan** [comment]
@Panajev thanks for bringing this up again. 

I agree this is something that should be implemented. There're a few capabilities that _fastlane_ is behind, I believe they should all be supported: `"NETWORK_CUSTOM_PROTOCOL", "COREMEDIA_HLS_LOW_LATENCY", "SYSTEM_EXTENSION_INSTALL", "USER_MANAGEMENT", "APPLE_ID_AUTH"`

I'm reopening this issue!

**Comment by markreynoso** [comment]
> @Panajev thanks for bringing this up again.
> 
> I agree this is something that should be implemented. There're a few capabilities that _fastlane_ is behind, I believe they should all be supported: `"NETWORK_CUSTOM_PROTOCOL", "COREMEDIA_HLS_LOW_LATENCY", "SYSTEM_EXTENSION_INSTALL", "USER_MANAGEMENT", "APPLE_ID_AUTH"`
> 
> I'm reopening this issue!

Yes!!

**Comment by joshdholtz** [comment]
Hello!

`produce` is currently using the legacy private API and it will be writing it soon to use the new App Store Connect API. I will make sure this is included in the rewrite 😊

**Comment by j3ski-passion** [comment]
Hi
@joshdholtz is there any ETA for the `produce` rewrite?

**Comment by joshdholtz** [comment]
Hoping to be complete in a few weeks! Turns out it's a whole thing 🤷‍♂️

**Comment by tmaly1980** [comment]
> Hoping to be complete in a few weeks! Turns out it's a whole thing 🤷‍♂️

Does this seem like it'll be ready by early January? Is there a separate ticket for the produce rewrite? Right now Apple is sending out emails saying they are enforcing 2FA without the ability to opt-out any longer, as of February 2021. Being able to use the new API for everything would be an ideal workaround for conforming to 2FA in a CI environment.

**Comment by jjochen** [comment]
any news on this @joshdholtz?

**Comment by joshdholtz** [comment]
@jjochen Its coming along a bit slower than planned 😬 Me and my wife had our first child so things not getting the time I used to or planned to have but I have not forgotten about this.

**Comment by MartinGonzalez** [comment]
@joshdholtz Congrats! :D

**Comment by joshdholtz** [comment]
@MartinGonzalez Thank you thank you!! 🥰

**Comment by jjochen** [comment]
> @jjochen Its coming along a bit slower than planned 😬 Me and my wife had our first child so things not getting the time I used to or planned to have but I have not forgotten about this.

@joshdholtz Congratulations indeed 😊

Someone mentioned [here](https://github.com/fastlane/fastlane/discussions/17889#discussioncomment-256065) that it's not possible at all. What do you think?

**Comment by joshdholtz** [comment]
Yeah, I just realized that a lot of those things are possible yet with the official App Store Connect API 😔 

My `produce` changes I'm working on are going to be slightly messed up in that basic usage will work with the API Key but doing things like this won't 😱 

I'm still going to try to add it in with Apple ID auth using private API 🤷‍♂️

**Comment by SiberianCMS** [comment]
> @SiberianCMS Did you ever get anywhere with this?

Sadly no, we are about to stop our ios publishing services due to the enforced 2FA, so we don't even search anymore for now.

**Comment by gholias** [comment]
We should not close this issue.

**Comment by gilsonleitesiqueirajunior** [comment]
Commenting to not auto-close. This feature is needed.

**Comment by gilsonleitesiqueirajunior** [comment]
This is still an issue.

**Comment by tmaly1980** [comment]
@joshdholtz Any luck yet?

**Comment by gilsonleitesiqueirajunior** [comment]
Please leave this open.

**Comment by rogerluan** [comment]
As @aurea-dristic said, it seems like this issue got resolved by #18853. If you face issues with `produce`, please open a new ticket 🙏
